### ISSUE TITLE
Conflict between several @types package

### ISSUE THREAD

**Issue by Keats** [Microsoft/TypeScript#9771]
**TypeScript Version:**  2.0.0

**Details**
I started a new project to try TS 2.0 and the `npm install @types/` way of getting typings and ran into a conflict issue.
```
$ npm install @types/react
$ npm install @types/react-router
```
React-router definition depends on `@types/react`.

**Expected behavior:**
Everything working fine

**Actual behavior:** 
```
ERROR in /home/vincent/Code/mobx-test/node_modules/@types/react-router/node_modules/@types/react/index.d.ts
(7,21): error TS2300: Duplicate identifier 'React'.

ERROR in /home/vincent/Code/mobx-test/node_modules/@types/react/index.d.ts
(7,21): error TS2300: Duplicate identifier 'React'.
```

**Comment by RyanCavanaugh** [comment]
@andy-ms can you take a look? We shouldn't be in this state for unversioned installs (note the installation of `react` in the `react-router`'s `node_modules` subdir, indicating a version conflict)

**Comment by andy-ms** [comment]
These two packages work for me when installed from scratch in either order.
React has `"version": "0.14.27"` and react-router has `"@types/react": "0.14.*"` as a dependency.

Could you please provide your `npm --version`? Versions of `npm` before version 3 did not flatten packages, so that could be the problem. If that's not it, could you tell us:

* The version in `node_modules/@types/react.package.json`
* The version and dependencies in `node_modules/@types/react-router/package.json`
* The version in `node_modules/@types/react-router/node_modules/@types/react/package.json`

**Comment by Keats** [comment]
I'm indeed still using it npm 2 (2.14.4)

**Comment by Tragetaschen** [comment]
I ran into this as well last week. I installed @types for jQuery and SignalR and this resulted in types for jQuery in `node_modules` of the most recent version and types for jQuery in `node_modules` below SignalR's folder of an older version.

Once I figured that out, I tried to update the packages and saw that the SignalR types had been updated on npm two hours earlier (but after I had installed them). It would be great if these types of dependencies on npm could be updated atomically, but I doubt that's possible.

The most annoying part was that running tsc on the command line or through `awesome-typescript-loader` didn't have problems with that, but Visual Studio showed these errors when building there.

**Comment by markbook2** [comment]
Here is a related problem and description:
File main.ts:
```
/// <reference path="jquery-1.10.d.ts"/>

import $ from "jquery"
import {myFunction} from "./my-module"

$('#id').click(function () {});
```

File my-module.ts:
```
/// <reference path="jquery-2.0.d.ts"/>

import $ from "jquery"

export function myFunction() {}
```

Files jquery-1.10.d.ts and jquery-2.0.d.ts are files with the same content, just for testing, but it can be slightly different for different version of the jQuery, and theirs content is latest version of the jquery.d.ts from the DefinitelyTyped (https://raw.githubusercontent.com/DefinitelyTyped/DefinitelyTyped/master/jquery/jquery.d.ts):
```
 md5sum jquery-1.10.d.ts jquery-2.0.d.ts
7ef0067d2faf006343aa95157f492867  jquery-1.10.d.ts
7ef0067d2faf006343aa95157f492867  jquery-2.0.d.ts
```

Now try to compile the main.ts with:
```
tsc --moduleResolution node -t ES5 main.ts
```

Expected result: **no errors**

Actual result:
```
jquery-2.0.d.ts(615,5): error TS2374: Duplicate string index signature.
jquery-2.0.d.ts(2844,5): error TS2374: Duplicate string index signature.
jquery-2.0.d.ts(2845,5): error TS2375: Duplicate number index signature.
jquery-2.0.d.ts(3218,5): error TS2300: Duplicate identifier 'export='.
my-module.ts(3,8): error TS1192: Module '"jquery"' has no default export.
jquery-1.10.d.ts(3218,5): error TS2300: Duplicate identifier 'export='.
main.ts(3,8): error TS1192: Module '"jquery"' has no default export.
```
I.e. you see batch of errors. It seems like tsc does not allow use different version of the .d.ts files with the same declarations for the included module/file. However I think it should not produce errors because the included module can use different version of the same library as shown above with file names - jquery-1.10.d.ts and jquery-2.0.d.ts.

tsc version:
```
tsc --version
Version 2.1.0-dev.20160805
```

**Comment by mhegazy** [comment]
The original issue has been fixed, but there are new issues, i have a fix queued up in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/10663

**Comment by mhegazy** [comment]
should be fixed in `@types/react-router@2.0.33`
